auth/login.go

```diff
@@ -0,0 +1,55 @@
+package auth
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/mesos/mesos-go/auth/callback"
+	"golang.org/x/net/context"
+)
+
+var (
+	NoLoginNameErr = errors.New("missing login name in context")
+)
+
+func Login(ctx context.Context, handler callback.Handler) <-chan error {
+	name, ok := LoginNameFrom(ctx)
+	if !ok {
+		return LoginError(NoLoginNameErr)
+	}
+	provider, ok := getAuthenticateeProvider(name)
+	if !ok {
+		return LoginError(fmt.Errorf("unidentified login name in context: %s", name))
+	}
+	return provider.Authenticate(ctx, handler)
+}
+
+func LoginError(err error) <-chan error {
+	ch := make(chan error, 1)
+	ch <- err
+	close(ch)
+	return ch
+}
+
+// unexported key type, avoids conflicts with other context-using packages
+type loginKeyType int
+
+const (
+	loginNameKey loginKeyType = iota // name of login provider to use, e.g. SASL
```

Comment:
loginName ~= loginProviderName. documentation definitely needed. Has nothing to do with SASL. The login provider thing was meant to abstract away the SASL/mechanism based interfaces. It's a small abstraction, but possibly useful. We can ditch it if you think it's going overboard. Are you familiar with Java's JAAS at all? The Login interface is somewhat inspired by JAAS LoginContext.

What's really ugly to me is how I return a `<-chan error` instead of just an `error` for Login and Authenticatee. Initially I was trying to emulate the "promise" API from the mesos implementation but I don't think it has a place in this Go-based API. If someone wants Login to run concurrently then they can just `go func() { ch <- auth.Login(...) }()` and use their own channel to process the result when it's convenient for them.